draft-ietf-jsonpath-base.md

```diff
@@ -176,30 +181,31 @@ Name:
 : The name in a name/value pair constituting a member.  (Also known as
   "key", "tag", or "label".)
 
+Array:
+: A JSON array as defined in {{-json}}.
+  Never used in its generic sense, e.g., for programming language objects.
+
 Element:
 : An item in an array.  (Also used with a distinct meaning in XML
   context for XML elements.)
 
+Index:
+: An unsigned number that identifies a specific element in an array.
```

Comment:
I disagree.  I prefer the simpler term.  Would like to hear from others whether they also want to fall back to "non-negative".

Reply:
I prefer non-negative for the reason @gregsdennis gave.

Reply:
I changed this as requested.

Reply:
Actually, I made this change because I don't really care too much, but I don't agree with it at all.

People are using unsigned type because they care about the application data type (which is unsigned), only rarely because the bits are then laid out in a specific way.

There is a long history of computer languages struggling with naming unsigned types.  FORTRAN only had INTEGER, which was signed.  In the 1970s, languages started to address the issue, but didn't find terminology that they could use from mathematics.  "Natural" is both (1..) and (0..) in common usage, so it is ambiguous.  A number of languages used "Cardinal" for unsigned integers.  C led the pack with going for "unsigned", and that is the term that has stuck.  But, incredibly, as late as the mid 1990s languages were designed that didn't have unsigned types (e.g., Java).

To me, requiring the clumsy term "non-negative" for unsigned is a bit like requiring to say "octet" for byte, except that the latter was actually a reasonable thing until about 1981 (when non-8-bit bytes became extinct).

Anyway, I'm not arguing for using "unsigned" in this PR, because I want to make progress, but, for the record, that's what it *should* be.

Reply:
@cabo I appreciate your flexibility.

A couple of things bothered me about using "unsigned":
* In JSONPath we are dealing with integers represented as strings rather than values of a numeric type.
* If we (and I hope we don't!) allowed indices such as "-0" and "+5", it would seem strange to describe these as "unsigned".

Reply:
@cabo I enjoyed your entertaining nostalgic review of integer types.

@glyn I am not aware of the fact, that in JSONPath we are dealing with integers represented as strings exclusively.

Reply:
We have two kinds of integers:

1. those coming from the JSON data item.  These are no longer in text form when we get to them; they are true numbers (no necessarily type-annotated to be integers).
2. those written in the JSONPath query.  I would expect that our processing model parses that query and feeds an AST to the process that already has true integers etc. as well.  But we haven't fully defined the processing model, so that is next.